Update Introduction and add Known Gaps (#42)
bifurcation/ietf-mimi-protocol: draft-ralston-mimi-protocol.md

# Introduction

> The More Instant Messaging Interoperability (MIMI) protocol enables providers of

Reviewer: ```suggestion
The More Instant Messaging Interoperability (MIMI) transport protocol enables providers of
```

end-to-end encrypted instant messaging to interoperate. As described in the MIMI
architecture {{?I-D.barnes-mimi-arch}}, group chats and direct messages are
described in terms of "rooms".  Each MIMI protocol room is hosted at a single
provider (the "hub" provider"), but allows users from different providers to
become participants in the room.  The hub provider maintains the policy and
participation for the room, authorizes messages, and is responsible for message

> ordering, the participation list, and policy of its rooms. Each provider also

Reviewer 2: Is there a reason we're distinguishing between "maintains the policy and participation for the room" and then later in the sentence "and is responsible for (...) the participation list, and policy of its rooms"?

Reviewer 3: I meant maintains the policy and participation as *enforcing* the policy and participation. It's also the source of truth ("responsible") for the "message ordering, the participation list, and policy of its rooms".

Feel free free to propose some alternate wording.

> become participants in the room.  The hub provider maintains the policy and
> participation for the room, authorizes messages, and is responsible for message
> ordering, the participation list, and policy of its rooms. Each provider also

Reviewer 2: Something like this? 
```suggestion
become participants in the room.  The hub provider is responsible for message ordering and distribution, as well as policy enforcement. It also keeps a copy of the room state, which includes the room's policy and participant list, which it can provide to new joiners. Each provider also
```
I wouldn't say it's the "source of truth", since the truth is in the MLS group state and the MLS commits. Saying it's the source of truth (to me) implies that it has direct control over it. It stores the group state, but the group state ultimately comes from the clients (even though the hub can reject commits it doesn't like).

Reviewer 2: The text looks good now, except that I don't quite get what the difference between "enforcing policy" and "authorizing messages" is. My understanding was that authorizing messages (or not) is part of policy enforcement.

> stores initial keying material and consent for its users (who may be offline).

Reviewer 3: ```suggestion
stores initial keying material and consent for its own users (who may be offline).
```

> stores initial keying material and consent for its users (who may be offline).

Reviewer 2: We should be more clear here what it means to "store consent". At least it's not clear to me.

Reviewer 3: if a user has granted, revoked, or rejected consent, that needs to get stored at their provider so the provider can act on behalf of the user if his or clients are offline.

Reviewer: in my opinion, we should descope consent for now. It needs a lot more thought as to how it works.

```suggestion
stores initial keying material for its own users (who may be offline).
```

Reviewer 4: This statement merely says that the providers are responsible for keeping track of the consent of their users, which should be non-controversial. This section / PR doesn't impose any specific mechanism for doing so.

Reviewer 2: Thanks for clarifying. I'd be okay to leave it as is for now and touch it again once we know more about how consent works.

This document describes the communication among different providers necessary to
support messaging application functionality, for example:

> * Sharing room policy

Reviewer 2: What does it mean to "share room policy"? Agreement between providers? Or participants?

In any case, the other two examples seem more natural in the context of "messaging application functionality", so I'd put them first.

Reviewer 3: I'd say "share room policy" is between participants and the hub.

Reviewer 2: It still sounds a bit weird to me, but I don't want to nitpick too much here. Let's leave it as is for now.

* Adding and removing participants in a room
* Exchanging secure messages

> In support of these functions, the protocol also has primitives to fetch initial
> keying material, fetch the current state of the underlying end-to-end encryption
> protocol for the room, and request, grant, and reject consent to communicate
> with users on other providers.

Reviewer: descoping consent: 
```suggestion
In support of these functions, the protocol also has primitives to fetch initial
keying material and fetch the current state of the underlying end-to-end encryption
protocol for the room.
```

Messages sent inside each room are end-to-end encrypted using the Messaging
Layer Security (MLS) protocol {{!RFC9420}}, and each room is associated with an
MLS group. MLS also ensures that clients in a room agree on the room policy and
participation.  MLS is integrated into MIMI in such a way as to ensure that a
client is joined to a room's MLS group only if the client's user is a

> participant in the group, and that all clients in the group agree on the state

Reviewer: ```suggestion
participant in the room, and that all clients in the group agree on the state
```

of the room (including, for example, the room's participant list).

## Known Gaps

In this version of the document, we have tried to capture enough concrete
functionality to enable basic application functionality, while defining enough

> of a protocol framework to indicate how other necessary functionality.  The

Reviewer 3: ```suggestion
of a protocol framework to indicate how to add other necessary functionality.  The
```

following functions are likely to be needed by the complete protocol, but are
not covered here:

Authorization policy:
: In this document, we assume that all participants in a room have equal
capability.  Actual messaging systems have authorization policies for which
clients can take which actions in a room.

Advanced join/leave flows:
: In this document, all adds / removes / joins / leaves are initiated from
within the group, since this aligns well with MLS.  Messaging application
support a variety of other flows, some of which this protocol will need to

> support.

Reviewer: This is a bit too vague: what flows specifically are we descoping from this version?

Reviewer 4: Richard wanted to make a join flow easy to understand by the WG, get consensus to make it a WG item, then add other join mechanisms. So in this version I think Richard wants to start with the "member adds non-member" flow. 

I'm fine (keen even) to add a second flow where a user adds a new client or a user already on a participant list (passively) does an external join, but only if we have strong consensus from the design team. 

@turt2live are you OK with that plan of action?

Author: Yeah, we're descoping everything besides the basic flow from this version.  The idea here is to capture the basic moving parts for one use cases, with the idea that the WG can then elaborate them to cover other use cases after adoption.

Consent:
: In this document, we assume that any required consent has already been
obtained, e.g., a user consenting to be added to a room by another user.  The
full protocol will need some mechanisms for establishing this consent.

Identifiers:
: Certain entities in the MIMI system need to be identified in the protocol.  In
this document, we define a notional syntax for identifiers, but a more
concrete one should be defined.

Abuse reporting:
: There is no mechanism in this document for reporting abusive behavior to a
messaging provider.

Identifier resolution:
: In some cases, the identifier used to initiate communications with a user
might be different from the identifier that should be used internally.  For
example, a user-visible handle might need to be mapped to a durable internal

> identifier.  This document provides no mechanism for such resolution.

Reviewer 2: If we do want to keep a list (I think another PR removes existing remarks about missing things), we should add authentication here. We do have basic MLS authentication, but we should also be able to verify cryptographically, that a given user belongs to a given provider.

> identifier.  This document provides no mechanism for such resolution.

Reviewer 4: ```suggestion
identifier.  This document provides no mechanism for such resolution.

```

> 

Reviewer 2: Maybe something like this?
```suggestion
Authentication
: While MLS provides basic message authentication, users should also be able to (cryptographically) tie the identity of other users to their respective providers. Further authentication such as tying clients to their users (or the user's other clients) may also be desirable.
```

# Introduction

The More Instant Messaging Interoperability (MIMI) transport protocol enables providers of
end-to-end encrypted instant messaging to interoperate. As described in the MIMI
architecture {{?I-D.barnes-mimi-arch}}, group chats and direct messages are
described in terms of "rooms".  Each MIMI protocol room is hosted at a single
provider (the "hub" provider"), but allows users from different providers to

> become participants in the room.  The hub provider maintains the policy and
> participation for the room, authorizes messages, and is responsible for message
> ordering, the participation list, and policy of its rooms. Each provider also

Reviewer 4: ```suggestion
become participants in the room. The hub provider is responsible for ordering
and distributing messages, enforcing policy, and authorizing messages. It also
keeps a copy of the room state, which includes the room policy and participant
list, which it can provide to new joiners. Each provider also
```

list, which it can provide to new joiners. Each provider also

> stores initial keying material and consent for its own users (who may be offline).

Author: ```suggestion
stores initial keying material for its own users (who may be offline).
```
